### PR TITLE
(PUP-9337) Update mode check to treat FullControl as insync for windows SYSTEM user

### PR DESCRIPTION
**Pull request by McdonaldSeanp**
This commit updates the mode calculations in puppet to treat "FullControl" as insync for the SYSTEM user in the following specific scenario:

1. the file resource does _not_ manage the group
2. the group set for the file is the "SYSTEM" (this is set automatically in windows when puppet runs as the SYSTEM user and creates a file)

This scenario occurs when a file resource does not manage the group on windows and the puppet run that creates the file is run as SYSTEM. Puppet runs will run as SYSTEM when puppet runs as a service, or during invocations from pxp-agent.

For more details on the underlying problems, see https://tickets.puppetlabs.com/browse/PUP-9337

### CODE REVIEW

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

File: lib/puppet/util/windows/security.rb [hunk 1]
```diff
@@ -247,11 +246,9 @@ def get_mode(path)
       # if owner and group the same, then user and group modes are the OR of both
       if sd.owner == sd.group
         mode |= ((mode & S_IRWXG) << 3) | ((mode & S_IRWXU) >> 3)
-        #puts "owner: #{sd.group}, 0x#{ace.mask.to_s(16)}, #{mode.to_s(8)}"
```
Comment (Iristyle) [inline lib/puppet/util/windows/security.rb]: These were actually in there for devs to uncomment while debugging... I think @joshcooper originally added them, but I've used them from time to time.

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

File: lib/puppet/provider/file/windows.rb [hunk 2]
```diff
@@ -71,7 +71,9 @@ def mode
 
   def mode=(value)
     begin
-      set_mode(value.to_i(8), resource[:path])
+      managing_group = resource[:group].nil? ? false : true
+      managing_owner = resource[:owner].nil? ? false : true
```
Comment (glennsarti) [inline lib/puppet/provider/file/windows.rb]: nit: `.nil?` already returns a boolean so no need to use a ternary here.

`managing_owner = !resource[:owner].nil?`

File: lib/puppet/provider/file/windows.rb [hunk 3]
```diff
@@ -86,6 +88,32 @@ def validate
...
   end
 
+  # munge the windows group permissions if the group is set to SYSTEM
+  #
+  # tl;dr: when SYSTEM user is the group and the resoure is not managing group then treat
+  # the resource as insync if System has FullControl access.
+  #
+  # On windows systems: files laid down when puppet is running as the SYSTEM user will have
+  # the SYSTEM user set as the "group" on them. In these situations we want puppet to ignore
+  # what the permissions should be as long as the system user has FullControl access.
```
Comment (glennsarti) [inline lib/puppet/provider/file/windows.rb]: Nit: Probably best to document what return types are expected here. From looking at it, we expect an integer or nil

What does a nil integer mean?

File: lib/puppet/provider/file/windows.rb [hunk 4]
```diff
@@ -86,6 +88,32 @@ def validate
...
+    if resource[:group].nil? && (group == Puppet::Util::Windows::SID::LocalSystem)
+      if (current.to_i(8) & group_fullcontrol) == group_fullcontrol
+        # Since the group is LocalSystem, and the permissions are FullControl,
+        # replace the value returned by group ownership with the value expected
+        # by 'should'. This will treat this specific situation as "insync"
+        current = ((current.to_i(8) & "707".to_i(8)) | (should[0].to_i(8) & "070".to_i(8))).to_s(8).rjust(4, '0')
+      else
+        # If the SYSTEM account does _not_ have FullControl in this scenario, we should
+        # force the resource out of sync no matter what.
+        Puppet.debug _("Not managing SYSTEM user and SYSTEM user does not have full control, treating mode as out of sync")
```
Comment (glennsarti) [inline lib/puppet/provider/file/windows.rb]: The message here is a bit misleading.
`Debug: Not managing SYSTEM user and SYSTEM user does not have full control, treating mode as out of sync`
Which resource?
Is this the owner or group?
I feel the msg needs more context to make it actionable for a user.

Also may need to add a note for Translators to NOT translate the word SYSTEM as that's a literal word

File: lib/puppet/type/file/mode.rb [hunk 5]
```diff
@@ -116,8 +116,16 @@ def dirmask(value)
     # If we're not following links and we're a link, then we just turn
     # off mode management entirely.
     def insync?(currentvalue)
+      if provider.respond_to?(:munge_windows_system_group)
+        munged_mode = provider.munge_windows_system_group(currentvalue, @should)
+        if munged_mode.nil?
```
Comment (glennsarti) [inline lib/puppet/type/file/mode.rb]: nit: you can avoid the branch here
```
return false if munged_mode.nil?
currentvalue = munged_mode
```

**Comment by glennsarti** [comment]
Also this is weird:

ACL Before:
``` powershell
C:\Source\tmp\puppet-pr7400\tmp>powershell "get-acl .\bad.txt | Select Owner, Group,AccessToString | fl"


Owner          : NT AUTHORITY\SYSTEM
Group          : NT AUTHORITY\SYSTEM
AccessToString : Everyone Allow  ReadAttributes, ReadPermissions, Synchronize
                 NT AUTHORITY\SYSTEM Allow  FullControl
```

Manifest:
``` puppet
file { 'C:/source/tmp/puppet-pr7400/tmp/bad.txt':
  ensure => present,
#  owner => 'SYSTEM',
  group => 'S-1-5-18',
  mode => '660'
}
```

Debug Log
`bundle exec puppet apply tmp\test.pp --debug`

```
...
Debug: Loaded state in 0.03 seconds
Debug: Loaded state in 0.04 seconds
Info: Applying configuration version '1551409221'
Debug: An attempt to set mode 660 on item C:/source/tmp/puppet-pr7400/tmp/bad.txt would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control
Warning: Attempting to set SYSTEM user access to less than Full Control rights. This may have unintented consequences for operations on this file
Notice: /Stage[main]/Main/File[C:/source/tmp/puppet-pr7400/tmp/bad.txt]/mode: mode changed '0770' to '0660'
Debug: /Stage[main]/Main/File[C:/source/tmp/puppet-pr7400/tmp/bad.txt]: The container Class[Main] will propagate my refresh event
Debug: Class[Main]: The container Stage[main] will propagate my refresh event
Debug: Finishing transaction 52450020
Debug: Storing state
...
```

First this
> This attempt has been corrected to Full Control

Then 
> Attempting to set SYSTEM user access to less than Full Control rights.

That's weird.  The debug line above says it was changing this to Full Control already, why am I getting a warning about less than Full Control

And then finally
> C:\Source\tmp\puppet-pr7400\tmp>powershell "get-acl .\bad.txt | Select Owner, Group,AccessToString | fl"
>
>
> Owner          : NT AUTHORITY\SYSTEM
> Group          : NT AUTHORITY\SYSTEM
> AccessToString : Everyone Allow  ReadAttributes, ReadPermissions, Synchronize
>                  NT AUTHORITY\SYSTEM Allow  FullControl
The _actual_ permissions put down are Full Control, so the Warning isn't needed?

**Comment by McdonaldSeanp** [comment]
@glennsarti I've updated the PR per your comments.

Specifically with the situation you highlighted last: It's doing both the FullControl force and warning because puppet is individually checking owner and group, which are both SYSTEM, but only the group is specifically managed. Since it performs the checks for user and group separately it's identifying group as a managed user and letting the user set SYSTEM to less than FullControl and emitting a warning. Then as it subsequently moves to user it identifies that user is SYSTEM but _not_ managed and is performing the force to FullControl. Since the ACL won't have different entries for the same user we end up with the FullControl setting from the owner.

We could try to remedy this by checking if _both_ owner and group are set, and try to find the logic around these scenarios.

But TBH I would rather not. The additional logic added here works, and trying to work around when both owner and group are SYSTEM (and only one is managed, or neither, or both, etc. etc.) seems unnecessary to me. It might look a little weird to the puppet user, but it seems to me this is an edge case. The logic required to work around the edge case seems too complex for little return.

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by glennsarti** [comment]
> but it seems to me this is an edge case

Owner and Group are set differently when run as a scheduled tasks e.g. Like we do in VRA/VRO.  This is what triggered the original PUP-9106 and PUP-6729 issues.  It's not as edge-case as you may think.

I'm beginning to feel we need some integration tests here to check round-trip-ability of various different combinations of owner, user, perms etc. to make sure we're not breaking anything intentionally

**Comment by glennsarti** [comment]
I've added a commit with integration tests for a bunch of combinations of owner, group, mode etc.

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

**Comment by McdonaldSeanp** [comment]
@glennsarti I've updated the PR from our discussion. 

Now when both owner and group are SYSTEM and one of owner/group is managed but the other is not puppet will emit debug messages but still end up with FullControl from the non-managed part. I think is what we want in that situation. It also will not emit a warning anymore, though maybe we should emit a specific warning in this situation instead of debug messages?

I've also fixed up the integration tests you wrote, see the last commit in the PR for details.

File: spec/integration/provider/file/windows_spec.rb [hunk 6]
```diff
@@ -2,158 +2,161 @@
...
-        # Idempotent. Should be no changed resources
-        expect(result.changed?.count).to eq 0
+      context_name = "With initial owner '#{testcase[:create_owner]}' and initial group '#{testcase[:create_owner]}'"
+      context_name += " and initial mode of '#{testcase[:create_mode]}'" unless testcase[:create_mode].nil?
+      context_name += " and a mode of '#{testcase[:manifest_mode]}' in the manifest"
+      context_name += " and an owner of '#{testcase[:manifest_owner]}' in the manifest" unless testcase[:manifest_owner].nil?
+      context_name += " and a group of '#{testcase[:manifest_group]}' in the manifest" unless testcase[:manifest_group].nil?
+
+      context context_name do
+        is_idempotent = testcase[:is_idempotent].nil? | testcase[:is_idempotent]
```
Comment (glennsarti) [inline spec/integration/provider/file/windows_spec.rb]: Whoops, typo. Should be `||`

File: spec/integration/provider/file/windows_spec.rb [hunk 7]
```diff
@@ -2,158 +2,161 @@
 require 'puppet_spec/compiler'
 require 'puppet_spec/files'
 
+# For some reason the provider test will not filter out on windows when using the
+# :if => Puppet.features.microsoft_windows? method of filtering the tests.
```
Comment (glennsarti) [inline spec/integration/provider/file/windows_spec.rb]: I suspect this may be due to some other tests mocking that call..

**Comment by McdonaldSeanp** [comment]
@glennsarti Updated to fix the typo you mentioned. Are there any other concerns with this PR? or do you think it's good to go?

File: lib/puppet/util/windows/security.rb [hunk 8]
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
       # It is possible to set SYSTEM with a mode other than Full Control (7) however this makes no sense and in practical terms
       # should not be done.  We can trap these instances and correct them before being applied.
       if (sd.owner == well_known_system_sid) && (owner_allow != FILE::FILE_ALL_ACCESS)
-        #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
-        Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
-        owner_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but group is unmanaged the control rights of system will be set to FUllControl by
+        # the unmanaged group, so there is no need for the warning
+        if managing_owner && !(isownergroup && !managing_group)
```
Comment (glennsarti) [inline lib/puppet/util/windows/security.rb]: `!(isownergroup && !managing_group)` is the same as `(!isownergroup || managing_group)`  Does that read better?

**Comment by glennsarti** [comment]
I've approved the PR.  Right now it's functional.  Now we're just quibbling over small things.

File: lib/puppet/util/windows/security.rb (hunk 8 above)
Comment (McdonaldSeanp) [reply lib/puppet/util/windows/security.rb]: Yeah, I think it might read a little better.

File: lib/puppet/provider/file/windows.rb [hunk 9]
```diff
@@ -86,6 +86,52 @@ def validate
...
+        if (current.to_i(8) & mode_part['fullcontrol']) == mode_part['fullcontrol']
+          # Since the group is LocalSystem, and the permissions are FullControl,
+          # replace the value returned with the value expected. This will treat
+          # this specific situation as "insync"
+          current = ( (current.to_i(8) & mode_part['remove_mask']) | mode_part['should_mask'] ).to_s(8).rjust(4, '0')
+        else
+          # If the SYSTEM account does _not_ have FullControl in this scenario, we should
+          # force the resource out of sync no matter what.
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.debug _("%{resource_name}: %{mode_part_type} set to SYSTEM. SYSTEM permissions cannot be set below FUllControl ('7')") % { resource_name: resource[:name], mode_part_type: mode_part['type']}
```
Comment (speedofdark) [inline lib/puppet/provider/file/windows.rb]: Change FUllControl to FullControl (uppercase U as second character).

File: lib/puppet/util/windows/security.rb [hunk 10]
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
       # It is possible to set SYSTEM with a mode other than Full Control (7) however this makes no sense and in practical terms
       # should not be done.  We can trap these instances and correct them before being applied.
       if (sd.owner == well_known_system_sid) && (owner_allow != FILE::FILE_ALL_ACCESS)
-        #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
-        Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
-        owner_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but group is unmanaged the control rights of system will be set to FUllControl by
```
Comment (speedofdark) [inline lib/puppet/util/windows/security.rb]: same FUllcontrol change to FullControl, please

File: lib/puppet/util/windows/security.rb [hunk 11]
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
...
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.warning _("Setting control rights for %{path} owner SYSTEM to less than Full Control rights. Setting SYSTEM rights to less than Full Control may have unintented consequences for operations on this file") % { path: path }
+        elsif managing_owner && isownergroup
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.debug _("%{path} owner and group both set to user SYSTEM, but group is not managed directly: SYSTEM user rights will be set to FullControl by group") % { path: path }
+        else
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
+          owner_allow = FILE::FILE_ALL_ACCESS
+        end
```
Comment (speedofdark) [inline lib/puppet/util/windows/security.rb]: I'm seeing a lot of differentiation between saying FullControl vs. Full Control. Do we need to worry about that?

File: lib/puppet/util/windows/security.rb [hunk 12]
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
...
       end
 
       if (sd.group == well_known_system_sid) && (group_allow != FILE::FILE_ALL_ACCESS)
-        #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
-        Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the group, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
-        group_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but owner is unmanaged the control rights of system will be set to FUllControl by
```
Comment (speedofdark) [inline lib/puppet/util/windows/security.rb]: FUllControl vs. FullControl, please.

File: lib/puppet/util/windows/security.rb (hunk 11 above)
Comment (McdonaldSeanp) [reply lib/puppet/util/windows/security.rb]: No, we are talking about the same thing.
